compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -994,6 +995,40 @@ static FailureOr<Type> nonWideningLinalgElementType(linalg::LinalgOp op) {
   return inputAndOutputElementTypes[0];
 }
 
+/// Utility to compute the tile sizes for AArch64 Neon/SVE.
+static void getMatmulAArch64NeonSVEVectorSizes(
+    func::FuncOp entryPointFn, linalg::LinalgOp op, int64_t vectorSize,
+    SmallVectorImpl<int64_t> &sizes, SmallVectorImpl<bool> &scalableSizeFlags) {
```

Comment:
Hm, I don't see anything specific to AArch64 in this method :) This makes me think that we are missing some more generic helpers.

I have a couple of more specific points:
* Everything starting from L1004 onward is adjusting **the 2nd parallel vector dim** (corresponding to the (N) dim in the input matrices). To me it sounds like we could introduce `adjust2ndVecParallelTile`? Or at least add comments to highlight that only the 2nd parallel dim is tweaked.
* ATM this method will only help when using something smaller than `i8` - the default "middle" dim is 16 and `16 * 8 = 128 bits`. Will it ever be needed for types >= `i8`? Perhaps limit this logic to cases where processing something < `i8`?

In general, IREE should probably adjust both parallel vector tile dims based the underlying element size? I've not thought about it much and you will have much more experience in this area.

Reply:
Yes, the idea is to start introducing more sophisticated heuristics for all the dimensions so this is just the start. We should add also logic to improve the other dimensions and that would probably be more target specific due to the different unrolling profiles each target has.

The fact that this only triggers for sub-byte types is also accidental. If we change the initial default value or introduce a multiplier, this would also trigger for i8 cases or even larger.

If you want I can try to apply this to all the targets for now but it's going to be very unlikely that we end up with a one-fits-all. Let me give that a try.

Reply:
I mostly want to make sure that  the heuristics that we use are clear and hard to violate :)  

This patch adjusts tiles sizes to make sure that we utilise full vectors - that's really key in this PR. But that's not that obvious from the code, e.g. `getMatmulAArch64NeonSVEVectorSizes`/`getDefaultMatmulVectorSizes` are fairly generic names. Perhaps there should be `adjustTileSizeToUseFullVector`  somewhere? Naming is hard.

> The fact that this only triggers for sub-byte types is also accidental. If we change the initial default value or introduce a multiplier, this would also trigger for i8 cases or even larger.

True, but if we change the initial value then we might be updating many other things as well, including these hooks. If we limit this to sub-byte then that would reflect the fact we have only really exercised this hook for sub-byte types.

> If you want I can try to apply this to all the targets for now but it's going to be very unlikely that we end up with a one-fits-all. 

This does feel as something that could benefit other targets. If you'd rather focus on one target, there could be a TODO saying "consider generalising to other targets". IMHO that would be a sufficient hint that this is potentially something generic, but has only been tested on AArch64. Right now it feels like there's a reason that this only applied to AArch64, but it's not clear "why".

Btw, I'm totally on board with this, this is great!

Reply:
Ok, I improved the names and doc. We can extract finer-grained utilities as we see fit. The added heuristic can't be restricted to the N dimension as the value for the N dimension will depend on the value for the M and K dimension (i.e., if we unroll N too much we shouldn't unroll M too much, etc.). IOW, the decision making should happen in one place looking at all the information available. We can make decisions on the N dimension without impacting the M dimension and vice-versa.

> True, but if we change the initial value then we might be updating many other things as well, including these hooks. If we limit this to sub-byte then that would reflect the fact we have only really exercised this hook for sub-byte types.

I don't think we should introduce type specific specializations as that would lead us back to hole we are starting to come out. This heuristic offers a generic and coherent output for any type combination but if you have any specific concerns on a specific case we can look at it. Again, it's not in its final state so it's expected to change in the near term.

I enabled this for other targets. Let's see how numbers look like.

Let me know if it makes more sense now!

Reply:
> IOW, the decision making should happen in one place looking at all the information available. We can make decisions on the N dimension without impacting the M dimension and vice-versa.

Agreed. The point that I was trying to make is that the current hook looks at (N) in isolation.

> I don't think we should introduce type specific specializations as that would lead us back to hole we are starting to come out. 

Agreed. Also, apologies, I should've been clearer - I meant adjusting the sizes based on the smallest element size as you do now.

> Let me know if it makes more sense now!

Yup, the intent is much clearer now, thanks!

> RISC-V ISel crashes... 

Thanks for checking - now we know why we can't generalise it :)